contrib/babelfishpg_tds/src/backend/tds/tdssecure.c

```diff
@@ -193,56 +193,71 @@ SslWrite(BIO * h, const char *buf, int size)
 static int
 SslHandShakeWrite(BIO * h, const char *buf, int size)
 {
-	StringInfoData str;
+	#define SSL_MAX_PACKET_SIZE 4096
+	StringInfoData	str;
 	char		tmp[2];
 	uint16_t	tsize;
-	int			res = 0;
+	int		res = 0;
+	int		total_written = 0;
+	int		packet_id = 0;
 
 	/* Nothing to write */
 	if (size < 0)
 		return size;
 
-	initStringInfo(&str);
-	appendStringInfoChar(&str, TDS_PRELOGIN);
-	appendStringInfoChar(&str, TDS_PACKET_HEADER_STATUS_EOM);
-	tsize = pg_hton16(size + TDS_PACKET_HEADER_SIZE);
-	memcpy(&tmp, (char *) &tsize, 2);
-
-	appendStringInfoChar(&str, tmp[0]);
-	appendStringInfoChar(&str, tmp[1]);
-	appendStringInfoChar(&str, 0x00);
-	appendStringInfoChar(&str, 0x00);
-	appendStringInfoChar(&str, 0x00);
-	appendStringInfoChar(&str, 0x00);
-
-	appendBinaryStringInfo(&str, buf, size);
-	buf = str.data;
-	size += TDS_PACKET_HEADER_SIZE;
-
-	/* Write the complete data */
-	while (res < size)
+	/* Process data in chunks of SSL_MAX_PACKET_SIZE */
+	while (size > 0)
```

Comment:
is there use case where size is 0? how do we handle that?

Reply:
We aren't really sure if the size can be zero. For sanity, we decided to keep the existing behaviour i.e. it will write 8 bytes of header if the given size is 0.